sdk/src/Services/DynamoDBv2/Custom/DataModel/ContextBuilder.cs

```diff
@@ -0,0 +1,153 @@
+
+        /// <summary>
+        /// Specifies if an <see cref="IAmazonDynamoDB"/> client is constructed or supplied externally
+        /// </summary>
+        private bool _ownClient = true;
+
+        /// <summary>
+        /// Specifies if the object was disposed to avoid redundant dispose calls
+        /// </summary>
+        private bool _disposed = false;
+
+        /// <summary>
+        /// Creates a builder object to construct a <see cref="DynamoDBContext"/>
+        /// </summary>
+        public DynamoDBContextBuilder()
+        {
+            _config = new DynamoDBContextConfig();
+            _config.DisableFetchingTableMetadata = true;
+        }
+
+        /// <inheritdoc/>
+        public IDynamoDBContextBuilder SetDynamoDBClient(IAmazonDynamoDB client)
+        {
+            if (_client is null)
+            {
+                _client = client;
+                _ownClient = false;
+            }
+            else
+            {
+                throw new AmazonDynamoDBException($"Cannot use both '{nameof(SetDynamoDBClient)}' and '{nameof(SetAWSRegion)}'. If you are supplying the DynamoDB client, set the AWS Region directly on the client.");
+            }
+
+            return this;
+        }
+
+        /// <inheritdoc/>
+        public IDynamoDBContextBuilder SetAWSRegion(RegionEndpoint region)
+        {
+            if (_client is null)
+            {
+                _client = new AmazonDynamoDBClient(region);
+                _ownClient = true;
+            }
+            else
+            {
+                throw new AmazonDynamoDBException($"Cannot use both '{nameof(SetDynamoDBClient)}' and '{nameof(SetAWSRegion)}'. If you are supplying the DynamoDB client, set the AWS Region directly on the client.");
+            }
+
+            return this;
+        }
+
+        /// <inheritdoc/>
+        public IDynamoDBContextBuilder ConfigureContext(Action<DynamoDBContextConfig> configure)
+        {
+            configure(_config);
+
+            return this;
+        }
+
+        /// <inheritdoc/>
+        public DynamoDBContext Build()
+        {
+            if (_client is null)
+            {
+                _client = new AmazonDynamoDBClient();
+                _ownClient = true;
+            }
+
+            _context = new DynamoDBContext(_client, _ownClient, _config);
+
+            return _context;
+        }
+
+        /// <summary>
+        /// Disposes of all managed and unmanaged resources.
+        /// </summary>
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        /// <summary>
+        /// Implements the Dispose pattern
+        /// </summary>
+        /// <param name="disposing">Whether this object is being disposed via a call to Dispose
+        /// or garbage collected.</param>
+        protected virtual void Dispose(bool disposing)
+        {
+            if (_disposed) return;
+
+            if (disposing && _context != null)
```

Comment:
I don't understand why we are disposing the context here. The builder gave the context back to the user and it is up to the user to decide when that should be disposed. Generally a builder is used once during the init of an application and the context is left around for the life time of the application. 

The builder should not be implementing `IDisposable` and it doesn't need a `_context` member variable. The context is a local variable creating in the `Build` method and returned to the user.

Reply:
I needed to implement `IDisposable` because I had `DynamoDBContext` and `IAmazonDBClient` as global members which required me to dispose them, but as you mentioned, I didn't need to be doing that.